code/game/machinery/shieldgen.dm

```diff
@@ -553,3 +553,668 @@
+		regen_boost += new_charger.charge_boost
+
+	calculate_regeneration()
+
+	max_strength_boost = initial(max_strength_boost)
+	for (var/obj/machinery/modular_shield/module/well/new_well in connected_modules)
+		max_strength_boost += new_well.strength_boost
+
+	calculate_max_strength()
+
+	radius_boost = initial(radius_boost)
+	for (var/obj/machinery/modular_shield/module/relay/new_relay in connected_modules)
+		radius_boost += new_relay.range_boost
+
+	calculate_radius()
+
+///Calculates the max radius the shield generator can support, modifiers go here
+/obj/machinery/modular_shield_generator/proc/calculate_radius()
+
+	max_radius = innate_radius + radius_boost
+
+	if(radius > max_radius)//the generator can no longer function at this capacity
+		deactivate_shields()
+		radius = max_radius
+
+///Calculates the max strength or health of the forcefield, modifiers go here
+/obj/machinery/modular_shield_generator/proc/calculate_max_strength()
+
+	max_strength = innate_strength + max_strength_boost
+	begin_processing()
+
+///Calculates the regeneration based on the status of the generator and boosts from network, modifiers go here
+/obj/machinery/modular_shield_generator/proc/calculate_regeneration()
+
+	max_regeneration = innate_regen + regen_boost
+
+	if(!active)
+		if(recovering)
+			current_regeneration = max_regeneration * 0.25
+			return
+		current_regeneration = max_regeneration
+		return
+
+	//we lose more than half the regeneration rate when generating a shield that is near the max
+	//radius that we can handle but if we generate a shield with a very small fraction
+	//of the max radius we can support we get a very small bonus multiplier
+	current_regeneration = (max_regeneration / (0.5 + (radius * 2)/max_radius))
+
+	if(!exterior_only)
+		current_regeneration *= 0.5
+
+///Reduces the strength of the shield based on the given integer
+/obj/machinery/modular_shield_generator/proc/shield_drain(damage_amount)
+	stored_strength -= damage_amount
+	begin_processing()
+	if (stored_strength < 5)
+		recovering = TRUE
+		deactivate_shields()
+		stored_strength = 0
+		update_icon_state()
+
+/obj/machinery/modular_shield_generator/process(seconds_per_tick)
+	stored_strength = min((stored_strength + (current_regeneration * seconds_per_tick)),max_strength)
+	if(stored_strength == max_strength)
+		if (recovering)
+			recovering = FALSE
+			calculate_regeneration()
+			update_icon_state()
+		end_processing() //we dont care about continuing to update the alpha, we want to show history of damage to show its unstable
+	if (active)
+		var/random_num = rand(1,deployed_shields.len)
+		var/obj/structure/emergency_shield/modular/random_shield = deployed_shields[random_num]
+		random_shield.alpha = max(255 * (stored_strength/max_strength), 40)
+
+
+
+//Start of other machines
+///The general code used for machines that want to connect to the network
+/obj/machinery/modular_shield/module
+
+	name = "Modular Shield Debugger" //Filler name and sprite for testing
```

Comment:
Are you gonna change the name for this?

Reply:
You arnt meant to be able to obtain this in game, its a basetype incase anyone wants to make their own subtype but didnt give it a name or sprite.

Do you want a different name for this?

Reply:
I read "for testing" and assumed you wanted it changed before merging. It's fine then